0000-ilp-ping-protocol/0000-ilp-ping-protocol.md

```diff
@@ -0,0 +1,78 @@
+* The **Known Condition** is the SHA256 hash of the Known Preimage, which when encoded using Base64 is `jAC8DGFPZPfh4AtZpXuvXFe2oRmpDVSvSJg2oT+bx34=`. This value is used in Unidirectional mode only.
+
+## Unidirectional Mode
+
+1. Initiator sends a ILP Prepare packet using the following details:
+   - **Destination**: Recipient's Interledger address.
+   - **Amount**: Any amount chosen by the Initiator.
+   - **Expiry**: Any appropriate expiry.
+   - **Condition**: The Known Condition, as defined in [Terminology](#terminology).
+   
+2. Upon receiving the Prepare packet, the Recipient MUST fulfill the payment using the following information:
+   - **Fulfillment**: The Known Fulfillment, as defined above.
+
+Note that a Recipient _MAY_ reject the payment if appropriate, for example due to an insufficient amount, invalid expiry, or for other factors.
+
+## Bidirectional Mode
+
+1. Initiator generates a new random 32-byte value to use as a Fulfilment (`F`) and generates the corresponding condition (`C`).
+2. Initiator sends an ILP Prepare packet to the Recipient with the following details:
+   - **Destination**: The Recipient's Interledger address.
+   - **Amount**: Any amount chosen by the Initiator.
+   - **Expiry**: An appropriate expiry that is long enough to allow for two round trips.
+   - **Condition**: The condition `C`.
+   - **Data**: A concatenated series of bytes with the following information:
+      - the bytes of the ASCII string `ECHOECHOECHOECHO`. In binary this value is `0x4543484F4543484F4543484F4543484F`.
+      - the byte `0x00`
+      - The ILP address of the Initiator (i.e., the return address) as an OER-encoded, variable length IA5 string.
+4. Upon receiving the Prepare packet, the Recipient identifies that it is a bidirectional Ping request by confirming the following:
+  1. The packet is addressed directly to itself (i.e. it has no tx or child suffix).
+  1. The packet has a payload conforming to the specification above (e.g., the byte following the `ECHOECHOECHOECHO` prefix in the payload is `0x00`).
+5. The Recipient DOES NOT immediately send an ILP Fulfill response. Instead the Recipient sends a new ILP `Prepare` packet addressed to the address parsed from the original original packet's payload. This new packet has following details:original 
+    - *Destination*: The Interledger address found in the original packet's data payload.
+    - *Amount*: Any amount chosen by the Initiator.
+    - *Expiry*: An expiry that is marginally smaller than the expiry on the original ILP Prepare.
+    - *Condition*: The same condition `C`.
+    - *Data*: A concatenated series of bytes with the following information:
+      - the bytes of the ASCII string `ECHOECHOECHOECHO` (i.e., `0x4543484F4543484F4543484F4543484F`).
+      - the byte `0x01`
+7. Upon receiving this packet, the Initiator identifies that it is a Pong by the fact that it is addressed to the address used in the original Ping request.
```

Comment:
Don't they identify that it's a pong based on the `0x01` byte in the data rather than the address? With the address, they just check that it's sent to their own address

Reply:
Hmmm...I was about to answer "yes", but I remembered [this comment](https://forum.interledger.org/t/document-the-ping-protocol/279/9) by @justmoon.

Reading that comment, it seems like an implementation should do the following:

1. Look to see if the destination address of an incoming packet _starts with_ its own ILP address. If so, then route the packet to some sort of "ping protocol" handler.
1. At some point, check for the "pong" byte as a verification.
1. Correlate the actual incoming address to an outgoing "ping" that was already sent.

E.g., Alice pings Bob using a packet containing the address `g.alice.abc123`. Bob will send a "pong" to `g.alice.abc123` and Alice can fulfill that.

All that said though, is this type of thing even needed? In the forum post, the comment that triggered the response was:

> It’s not clear how Alice could differentiate a Pong from other packets right now as she uses her address with no additional suffix.

Does Alice need to differentiate here?